### ISSUE TITLE
Save Configuration on Exit: Possible Suggestion

### ISSUE THREAD

**Issue by elliott94** [nvaccess/nvda#5523]
Currently, if the user has chosen the option for NVDA to automatically save the configuration on exit, the nvda.ini file appears to get updated no matter if any settings have changed within that particular session. Whilst in theory this is how the setting is supposed to work, it can result in a potentially strange situation whereby if a user manually copies an alternative configuration file whilst NVDA is active (for example if they want to use settings from a custom install), this file will always get overwritten by default - I experienced this a while ago, and it took me ages to figure out what the problem was!

Taking the above into account, would it in theory be possible to only instruct NVDA to update the config file on disk if the user makes any changes within the running copy? This would allow an alternative configuration file to be placed in its directory without the chance of experiencing the same overwriting issue.

**Comment by leonardder** [comment]
When you manually override a nvda.ini, choosing for "revert to saved configuration" from the NVDA menu should be a nice alternative to prevent your issue, I think.

**Comment by elliott94** [comment]
I agree - for the average user though, this solution may not be the most obvious. That said, I realise that manually copying configuration data may not be the first thing a user does after a clean install! :)

**Comment by josephsl** [comment]
Hi,

 

This may not be the case: when you choose to revert configuration, whichever data that is stored on nvda.ini will be loaded, thus introducing the alternate configuration file. I’ll need to log into Github to comment on the original post, as I can see some disadvantages…

 

Thanks.

**Comment by josephsl** [comment]
Hi,

Although in theory this is possible and desirable on some systems such as systems using solid-state drives (SSD), I see potential problems;

## Currently...

1. When you quit NVDA, NVDA will first check if it can save user configuration upon exit, and if so, it'll call config.conf.save. This uses ConfigObj.write method to flush changes to disk.
2. When you tell it to revert from a saved configuration, NvDA will load it from the predefined nvda.ini file and doesn't care about the content (internally, it calls the constructor for ConfigObj). See my notes for details on why it does not fit the suggestion.

## Ideally...

1. NVDA will keep a copy of the original configuration and will prevent anyone (including NVDA itself) to modify this. this could be achieved by using tuples or extending the ConfigManager.
2. When quitting NVDA, NVDA will compare the runtime configuration versus the configuration cache stored earlier, and will flush the changes to disk if there is even one change. For most cases, this will be skipped.

## Advantages

* This may help for those using SSD's, in that we can help extend life of an SSD by taking precautions to not write to nvda.ini when not needed.
* Allows configuration reversion to occur straight from RAM as opposed to reading the file from disk, as a read-only copy of the config cache from disk will be available at all times.

## Possible issues with the suggestion

* What if someone does copy an alternate configuration to nvda.ini and tells NVDA to "revert" configuration? This will result in the alternate configuration coming to life and it is considered undesirable: trying to use a synthesizer that doesn't exist, loading wrong settings for normal configuration and so on.
* One should not mess with nvda.ini, as it controls NVDA's behavior.
* There is a command line to specify the alternate configuration file.
* Having a RAM cache isn't the ideal solution if it will be stored on data structures that could be tampered easily such as lists and dictionaries. A special case is if one performs factory reset, in which case the RAM cache becomes invalid (one can then revert this change) and will make the alternate configuration invalid as well, negating this suggestion.
* Care must be taken to ensure that this mechanism does not disturb configuration profiles, and if RAM cache is adopted, store configuration profiles along with normal configuration (Jamie knows more about this, as he worked on #667 throughout summer of 2013).

Thanks.

**Comment by jcsteh** [comment]
I understand the desire to have this. However, the "average user" won't be copying configuration files around; it's certainly not a documented (and thus officially supported) feature. I think the extra code to support this isn't worth the gain, especially given that there's a simple workaround. It might, hwever, be useful to have in some sort of advanced user FAQ.

**Comment by elliott94** [comment]
I'm more than happy to write some initial documentation on this to provide the suggested method of carrying out this task; if this would be in any way useful, where would you suggest it should go in the User Guide?

**Comment by jcsteh** [comment]
I don't know that it fits in the User Guide at all; we tend to only 
document officially supported stuff in the User Guide. I think a wiki 
page or similar is going to be best.
